qiskit_experiments/analysis/curve_analysis.py

```diff
@@ -0,0 +1,708 @@
+        y_sigmas = np.asarray(y_sigmas, dtype=float)
+
+        # Find series (invalid data is labeled as -1)
+        series = -1 * np.ones(x_values.size, dtype=int)
+        for idx, series_def in enumerate(self.__series__):
+            data_index = np.asarray(
+                [_is_target_series(datum, **series_def.filter_kwargs) for datum in data], dtype=bool
+            )
+            series[data_index] = idx
+
+        return x_values, y_values, y_sigmas, series
+
+    def _format_fit_options(self, **fitter_options) -> Dict[str, Any]:
+        """Format fitting option args to dictionary of parameter names.
+
+        Args:
+            fitter_options: Fit options generated by `self._setup_fitting`.
+
+        Returns:
+            Formatted fit options.
+
+        Raises:
+            QiskitError:
+                - When fit functions have different signature.
+            KeyError:
+                - When fit option is dictionary but key doesn't match with parameter names.
+                - When initial guesses are not provided.
+            ValueError:
+                - When fit option is array but length doesn't match with parameter number.
+        """
+        # check fit function signatures
+        fsigs = set()
+        for series_def in self.__series__:
+            fsigs.add(inspect.signature(series_def.fit_func))
+        if len(fsigs) > 1:
+            raise QiskitError(
+                "Fit functions specified in the series definition have "
+                "different function signature. They should receive "
+                "the same parameter set for multi-objective function fit."
+            )
+        fit_params = list(list(fsigs)[0].parameters.keys())[1:]
+
+        # Validate dictionaly keys
+        def _check_keys(parameter_name):
+            named_values = fitter_options[parameter_name]
+            if not named_values.keys() == set(fit_params):
+                raise KeyError(
+                    f"Fitting option {parameter_name} doesn't have the "
+                    f"expected parameter names {','.join(fit_params)}."
+                )
+
+        # Convert array into dictionary
+        def _dictionarize(parameter_name):
+            parameter_array = fitter_options[parameter_name]
+            if len(parameter_array) != len(fit_params):
+                raise ValueError(
+                    f"Value length of fitting option {parameter_name} doesn't "
+                    "match with the length of expected parameters. "
+                    f"{len(parameter_array)} != {len(fit_params)}."
+                )
+            return dict(zip(fit_params, parameter_array))
+
+        if "p0" in fitter_options:
+            if isinstance(fitter_options["p0"], dict):
+                _check_keys("p0")
+            else:
+                fitter_options["p0"] = _dictionarize("p0")
+        else:
+            raise KeyError("Initial guess p0 is not provided to the fitting options.")
```

Comment:
How does a class write a function to guess `p0?` Should this be a method of this curve fit class?

Reply:
There is `_setup_fitting` method to define a list of or single fit condition. If multiple conditions are created by subclass, this base class tries to fit for all given fit conditions and chose the best one based on the chisq value.